mlflow/R/mlflow/tests/testthat/test-rest.R

```diff
@@ -49,3 +52,47 @@ test_that("insecure is used", {
 
   expect_equal(rest_config$config, httr::config(ssl_verifypeer = 0, ssl_verifyhost = 0))
 })
+
+
+test_that("429s are retried", {
+  next_id <<- 1
+  client <- mlflow:::mlflow_client("local")
+  new_response <- function(status_code) {
+    structure(
+      list(status_code = status_code,
+           content = charToRaw('{"text":"text"}'),
+           headers = list(`Content-Type` = "raw")
+      ), class = "response"
+    )
+  }
+  responses <- list(new_response(429), new_response(429), new_response(200))
+  with_mock(.env = "httr", GET = function(...) {
+    res <- responses[[next_id]]
+    next_id <<- next_id + 1
+    res
+  }, {
+    tryCatch({
+      mlflow_rest(client = client, max_rate_limit_interval = 0)
+      stop("The rest call should have returned 429 and the function should have thrown.")
+    }, error = function(cnd) {
+      # pass
+      TRUE
+    })
+    x <- mlflow_rest(client = client, max_rate_limit_interval = 2)
+    expect_equal(x$text, "text")
+    next_id <<- 1
+    x <- mlflow_rest(client = client, max_rate_limit_interval = 2)
```

Comment:
Nit: Looks like we test the exact same behavior twice? i.e.

```
    x <- mlflow_rest(client = client, max_rate_limit_interval = 2)
    expect_equal(x$text, "text")
    next_id <<- 1
```

is repeated

Reply:
There is a slight difference, the first call does not reset nex_id so only the first call returns 429, the second call returns 200. Not sure if we need this as a separate test case though.